Soft delete reviews & responses when a release is deleted

GCHQDeveloper36 opened the pull request:
Currently, when a release is deleted from a model entry, the reviews & responses are left in the system. In particular, this leads to the reviews page breaking since it can't find the related semver/model combination.

This change finds the reviews associated with a release, and any responses associated with that review, and soft deletes them before continuing to soft delete the release.

There would be some benefit to wrapping these in a transaction in the future once #1237 is merged.

ARADDCC002 commented:
I'll get time to properly review this next week, but I wonder if this needs a migration script to cover any preexisting orphaned reviews and responses? 

We've had an issue in the past where the side bar navigation component occasionally doesn't render, and I wonder if it's because of dodgy reviews (we fetch user reviews to determine the number for that icon over the `/reviews` button to display how many active reviews you have), so there's a very slim (but possible!) chance it might help resolve that.

GCHQDeveloper36 commented:
> I'll get time to properly review this next week, but I wonder if this needs a migration script to cover any preexisting orphaned reviews and responses?
> 
> We've had an issue in the past where the side bar navigation component occasionally doesn't render, and I wonder if it's because of dodgy reviews (we fetch user reviews to determine the number for that icon over the `/reviews` button to display how many active reviews you have), so there's a very slim (but possible!) chance it might help resolve that.

I've added a migration script and moved the code around a little. Would probably benefit from running this on a few dev instances to make sure it's picking up the right releases/reviews/responses.

ARADDCC002 commented on backend/src/services/response.ts:
```diff
@@ -61,6 +62,22 @@ export async function updateResponse(user: UserInterface, responseId: string, co
   return response
 }
 
+export async function removeResponses(user: UserInterface, parentIds: string[]) {
```
`user` doesn't look like it's being used

ARADDCC002 commented on backend/src/migrations/019_2426_remove_orphaned_reviews_responses.ts:
```diff
@@ -0,0 +1,41 @@
+
+export async function up() {
+  // Find all releases have already been marked as deleted
+  const deletedReleases = await ReleaseModel.find({ deleted: true }).lean()
```
We should remove `lean` as it's the semver is failing validation when this object is returned as a POJO. We also need to consider access request reviews as well as releases

JR40159 commented on backend/src/services/review.ts:
```diff
@@ -113,6 +113,31 @@ export async function removeAccessRequestReviews(accessRequestId: string) {
   return deletions
 }
 
+export async function removeReleaseReviews(user: UserInterface, modelId: string, semver: string): Promise<ReviewDoc[]> {
```
Could you include some unit tests to cover this?
